llvm/lib/Transforms/Scalar/LoopInterchange.cpp

```diff
@@ -187,11 +187,19 @@ static void interChangeDependencies(CharMatrix &DepMatrix, unsigned FromIndx,
 // if the direction matrix, after the same permutation is applied to its
 // columns, has no ">" direction as the leftmost non-"=" direction in any row.
 static bool isLexicographicallyPositive(std::vector<char> &DV) {
+  bool HasScalar = false;
+  bool FirstPosScalar = true;
```

Comment:
I don't fully understand what `FirstPosScalar` means. This legality check seems to allow the permutation of `[S <]` to `[< S]`. Is this transformation legal? Or can such a case not happen? I tried some loops that have a row of `[S <]`, and as far as these cases, the legality check rejects the interchange because of another row like `[S =]`.

Reply:
I would second this. It looks to me that this patch allows some specific pattern to be interchanged without a schematic reasoning.